subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java

```diff
@@ -82,108 +85,132 @@ public <V> CrossBuildInMemoryCache<Class<?>, V> newClassMap() {
-                valuesForThisSession.clear();
-            }
+            retainValuesFromCurrentSession(valuesForThisSession.values().stream().map(Lazy::get));
+            valuesForThisSession.clear();
         }
 
         @Override
         public void clear() {
-            synchronized (lock) {
-                valuesForThisSession.clear();
-                discardRetainedValues();
-            }
+            valuesForThisSession.clear();
+            discardRetainedValues();
         }
 
-        protected abstract void retainValuesFromCurrentSession(Collection<V> values);
+        protected abstract void retainValuesFromCurrentSession(Stream<V> values);
 
         protected abstract void discardRetainedValues();
 
+        /**
+         * Must be thread-safe.
+         */
         protected abstract void retainValue(K key, V v);
 
+        /**
+         * Must be thread-safe.
+         */
         @Nullable
         protected abstract V maybeGetRetainedValue(K key);
 
         @Nullable
         @Override
         public V getIfPresent(K key) {
-            synchronized (lock) {
-                return getIfPresentWithoutLock(key);
-            }
+            Lazy<V> present = valuesForThisSession
+                .computeIfAbsent(key, k -> {
+                    V retained = maybeGetRetainedValue(k);
+                    return retained != null
+                        ? Lazy.fixed(retained)
+                        : null;
+                });
+            return present != null
+                ? present.get()
+                : null;
         }
 
+        /**
+         * Factory must be thread-safe and must not rely on thread-local state as it might
+         * be executed from a different thread than the one that submitted it.
+         */
         @Override
         public V get(K key, Function<? super K, ? extends V> factory) {
-            synchronized (lock) {
-                V v = getIfPresentWithoutLock(key);
-                if (v != null) {
-                    return v;
-                }
-
-                // TODO - do not hold lock while computing value
-                v = factory.apply(key);
-
-                retainValue(key, v);
-
-                // Retain strong reference
-                valuesForThisSession.put(key, v);
-
-                return v;
-            }
+            return valuesForThisSession.computeIfAbsent(key, k -> {
+                V retained = maybeGetRetainedValue(k);
+                return retained != null
+                    ? Lazy.fixed(retained)
+                    : Lazy.locking().of(() -> produceAndRetain(factory, k));
+            }).get();
         }
 
         @Override
         public void put(K key, V value) {
-            synchronized (lock) {
-                retainValue(key, value);
-                valuesForThisSession.put(key, value);
-            }
+            // Update doesn't need to be atomic since all retained values are equivalent
```

Comment:
Not sure I fully understand this comment. Equivalent as in one retained value will always `.equals` another? 

If so, how is this guaranteed?

Reply:
It is an assumption that factories produce equivalent values from keys. They are equivalent in their usefulness of being retained to be reused in the next session.